Html5 driver implementation of `NdtSingleTestResult`

rowanv opened the pull request:
Implemented `NdtSingleTestResult` -- `NdtResult` now stores `NdtSingleTestResult` objects, rather than storing the values directly. Also converted values so they are in Mb/s where necessary.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/ndt-e2e-clientworker/6)
<!-- Reviewable:end -->

mtlynch commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[client_wrapper/client_wrapper.py, line 34 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KDxlRVBg-efr47q0vlV:-KDxlRVCaI_B_QHlb_YJ:-541883976) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/149d4c22b54585cb0af51a5f626358510a2939d8/client_wrapper/client_wrapper.py#L34)):*
This looks like it may be debugging code that was accidentally left behind.

---

*[client_wrapper/html5_driver.py, line 216 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KDxmCSg1wJZ5LzVdTzn:-KDxmCSg1wJZ5LzVdTzo:-1058204039) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/149d4c22b54585cb0af51a5f626358510a2939d8/client_wrapper/html5_driver.py#L216)):*
Throughput should be a float rather than a string.

I believe both c2s and s2c can be displayed in Mbps or Kbps. Can we have a single parsing function for both like `_parse_throughput` so that the calls look like:

```
c2s_throughput_raw = driver.find_element_by_id('upload-speed').text
try:
  result.c2s_result.throughput = _parse_throughput(c2s_throughput_raw)
except ResultParseError as ex:
  results.errors.append(results.TestError(
    datetime.datetime.now(pytz.utc),
    'failed to parse c2s throughput: %s' % ex))
```

---

*[tests/test_html5_driver.py, line 283 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KDxoK1_UV-y60LLVTiT:-KDxoK1aCJvjV6cumHc_:-917771359) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/149d4c22b54585cb0af51a5f626358510a2939d8/tests/test_html5_driver.py#L283)):*
Double quotes

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6)*
<!-- Sent from Reviewable.io -->

rowanv commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[client_wrapper/html5_driver.py, line 216 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KDxmCSg1wJZ5LzVdTzn:-KDxsRMKdQlNNetMVNQM:-918266865) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/149d4c22b54585cb0af51a5f626358510a2939d8/client_wrapper/html5_driver.py#L216)):*
In that case, should I change all of the numerical values to be floats rather than strings? Selenium reads in the values from web elements as strings, so I was just changing it to be a string for consistency. 

I'm a bit unclear on the `_parse_throughput` function -- it seems that we want a single function that converts both values. However, if c2s is originally in kb/s in the web app, but s2c is originally in Mb/s in the web app, then it seems that one of the values does not need to be converted?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[client_wrapper/html5_driver.py, line 216 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KDxmCSg1wJZ5LzVdTzn:-KDxyzDDYbZgC7xRl7U-:-786317883) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/149d4c22b54585cb0af51a5f626358510a2939d8/client_wrapper/html5_driver.py#L216)):*
Yes, values as floats please.

c2s isn't guaranteed to be in kb/s. The units for both c2s and s2c change depending on the value displayed:

![throttled-ndt-connection.png](https://files.reviewableusercontent.io/images/reviewable/7783288/Hw66ukOU-3kZ2uFIsd7UZf0K/throttled-ndt-connection.png)

https://github.com/ndt-project/ndt/blob/master/HTML5-frontend/script.js#L466

We should support kb/s, Mb/s, and Gb/s.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6)*
<!-- Sent from Reviewable.io -->

rowanv commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[client_wrapper/client_wrapper.py, line 34 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KDxlRVBg-efr47q0vlV:-KE6WYf_2LL5Wc0uwBlt:804027418) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/149d4c22b54585cb0af51a5f626358510a2939d8/client_wrapper/client_wrapper.py#L34)):*
whoops, fixed!

---

*[client_wrapper/html5_driver.py, line 216 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KDxmCSg1wJZ5LzVdTzn:-KE6Xtkn7fb3js5CRr1b:-703604411) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/149d4c22b54585cb0af51a5f626358510a2939d8/client_wrapper/html5_driver.py#L216)):*
Ohhh good deal, that makes sense. Sounds good, went ahead and added the conversion logic for all 3 unit types.

---

*[tests/test_html5_driver.py, line 283 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KDxoK1_UV-y60LLVTiT:-KE6Wffur1eYlsJrXrue:-2003896356) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/149d4c22b54585cb0af51a5f626358510a2939d8/tests/test_html5_driver.py#L283)):*
Thanks, fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[client_wrapper/html5_driver.py, line 238 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6qVRgPe0uqIAFhQ2s:-KE6qVRgPe0uqIAFhQ2t:-1637899173) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L238)):*
It looks like think "metric" is a carryover from when this function was covering "latency" also. Can we replace "metric" with "throughput"?

---

*[client_wrapper/html5_driver.py, line 245 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6qoivehJGR-U6ymN3:-KE6qoivehJGR-U6ymN4:-427415451) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L245)):*
We don't use anything in the `result` instance here except for the `errors` field, so suggest making this parameter `errors` and the caller calls it with `_parse_throughput(result.errors, ...`

---

*[client_wrapper/html5_driver.py, line 253 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6qKf6L5_IX0egXuGr:-KE6qKf6L5_IX0egXuGs:1531803085) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L253)):*
remove "bool"

---

*[client_wrapper/html5_driver.py, line 256 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6rlDj_vcwi4Sr-GJX:-KE6rlDj_vcwi4Sr-GJY:-55324912) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L256)):*
We're parsing the string to a float, throwing the result away, and then parsing it over again in the last clause of this function. Can we save the result to avoid repeated parsing?

---

*[client_wrapper/html5_driver.py, line 270 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6rA2fhhheQRDu3F2Y:-KE6rA2fhhheQRDu3F2Z:-2009639724) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L270)):*
We should check for 'Mb/s' explicitly and add an error, then return `None` if it doesn't match any of the expected units. Otherwise if there's something like 'Tb/s' we'd assume it's Mb/s.

---

*[client_wrapper/html5_driver.py, line 274 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6sSZB7VWva4IYTGqg:-KE6sSZB7VWva4IYTGqh:-1487460558) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L274)):*
Need blank lines between one line summary, full summary, args, and returns sections.
https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments

---

*[client_wrapper/html5_driver.py, line 282 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6sCetPaUmyTEfOi6G:-KE6sCeuvg9sVIIeZ568:-1319365427) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L282)):*
These semantics are a bit confusing because it's unintuitive for the caller to figure out if the call failed. Suggest returning `None` on parse failure.

---

*[client_wrapper/html5_driver.py, line 291 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6t-povmdv2pz1A748:-KE6t-povmdv2pz1A749:-58114932) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L291)):*
This code looks identical to what's in `_parse_throughput`. Can we refactor out the common code to avoid duplication?

---

*[tests/test_html5_driver.py, line 304 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6u1Fne4yzsJU29VtL:-KE6u1Fne4yzsJU29VtM:925473410) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/tests/test_html5_driver.py#L304)):*
Instead of `NewWebElement` we can do `mock.Mock(text='Gb/s')` (ditto for the rest of the `NewWebElement` usages, so we can delete the definition of the `NewWebElement` class.

---

*[tests/test_html5_driver.py, line 306 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6uGSB0AcmSFeFmVPr:-KE6uGSB0AcmSFeFmVPs:694239026) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/tests/test_html5_driver.py#L306)):*
For the sake of realistic tests, we should use strings rather than ints (`'72'` instead of `72`)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6)*
<!-- Sent from Reviewable.io -->

rowanv commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[client_wrapper/html5_driver.py, line 238 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6qVRgPe0uqIAFhQ2s:-KE78YJ88es3gOpe-dOJ:-1962606206) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L238)):*
Sounds good, fixed.

---

*[client_wrapper/html5_driver.py, line 245 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6qoivehJGR-U6ymN3:-KE74E8IEkmBfV8YDuK2:-1962606206) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L245)):*
Sounds good, fixed.

---

*[client_wrapper/html5_driver.py, line 253 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6qKf6L5_IX0egXuGr:-KE75B75CuQToPJYOwI2:-1620474753) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L253)):*
fixed.

---

*[client_wrapper/html5_driver.py, line 256 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6rlDj_vcwi4Sr-GJX:-KE75V-drH9evGSvF9Zt:-1962606206) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L256)):*
Sounds good, fixed.

---

*[client_wrapper/html5_driver.py, line 270 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6rA2fhhheQRDu3F2Y:-KE7IBPpm-Y2qDFIF7xh:855492723) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L270)):*
Sounds good, fixed. Required reworking a good number of the unit tests as well. 

---

*[client_wrapper/html5_driver.py, line 274 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6sSZB7VWva4IYTGqg:-KE75xVZhmNcKsx1o0Sp:-2003896356) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L274)):*
Thanks, fixed.

---

*[client_wrapper/html5_driver.py, line 282 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6sCetPaUmyTEfOi6G:-KE76J0mf7kDPVDZfiYY:-1962606206) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L282)):*
Sounds good, fixed.

---

*[client_wrapper/html5_driver.py, line 291 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6t-povmdv2pz1A748:-KE77i_bAxtDsoyh-8_O:1079064994) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/client_wrapper/html5_driver.py#L291)):*
sounds good, fixed.

---

*[tests/test_html5_driver.py, line 304 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6u1Fne4yzsJU29VtL:-KE7CmClcWv4Bufa761G:-1962606206) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/tests/test_html5_driver.py#L304)):*
Sounds good, fixed.

---

*[tests/test_html5_driver.py, line 306 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-KE6uGSB0AcmSFeFmVPr:-KE7BpRTmclvWJspCtEx:-1962606206) ([raw file](https://github.com/m-lab/ndt-e2e-clientworker/blob/b5e18237a37f021a6aa6482305f47fd80b1be975/tests/test_html5_driver.py#L306)):*
Sounds good, fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt-e2e-clientworker/6#-:-KE7NYaES3L7NXsCheZ9:1417088721)*
<!-- Sent from Reviewable.io -->
